### ISSUE TITLE
Add capability to redeploy helm charts without failure

### ISSUE THREAD

**Issue by carterwilliamson** [pulumi/pulumi-kubernetes#215]
**Problem Statement**
Helm deployments cannot be deployed idempotently.

**Expected Behavior**
Re-running `pulumi up` on a stack with Helm charts should not fail if the chart already exists on the cluster. Alternatively this could be a flag/option.

**Current Behavior**
Re-running `pulumi up` generates errors like this...
`error: Plan apply failed: 1 error occurred: * secrets "alertmanager-kube-prometheus" already exists`

**Notes**
Helm has this capability using `helm upgrade --install`
This will fail if the chart exists in a FAILED release state. To fix this use the `--force` flag.

It would be nice to have upgrade/force as options in Pulumi helm to mimic this behavior.

**Comment by hausdorff** [comment]
`pulumi up` should _never_ fail if it successfully created a resource before. In other words, this _should_ work if there are not two different `Secret`s that are named the same thing (_e.g._, if two charts try to create a `Secret` with the same name).

Unfortunately, the three Helm charts @carterwilliamson is using appear to have no name collision between any `Secret`s, so I think we can rule that possibility out. It now seems certain that the Kubernetes provider is somehow forgetting about a resource it had previously created. (cc @joeduffy, who proposed that this might be a naming problem).

I have some ideas about what could have caused such a thing:

* Delete bug. If the `new k8s.helm.v2.Chart` call was renamed (say from `"coreos/kube-prometheus"` -> `"kube-prometheus"`), Pulumi would plan a delete-after-replacement. But, because Pulumi can't auto-name Helm resources, the "replacement" step would try to create a "new" `Secret` with the same name, which would fail because a `Secret` with that name already exists.
* ctrl-c. There could be a race condition somewhere that causes us to accidentally orphan resources.
* There is some kind of error in the partial error logic. (This does not seem likely.)

Unfortunately, after a couple hours' effort, I can't seem to reproduce this, so I can't say for sure which of these it is. Talking to @carterwilliamson, I understood the steps to encounter the error:

* Boot up AKS
* Boot up kube-prometheus v0.0.105, which fails, because you need the Prometheus operator.
* Boot up prometheus-operator chart v0.0.29
* Try again to boot up kube-prometheus, which this time fails because it claims the secret already exists.

@carterwilliamson: if something here is wrong, please correct me. If that looks right, then it would be very helpful if you could see if you could get it to fail reliably again, and then we can try to figure it out together.

**Comment by joeduffy** [comment]
Rereading @carterwilliamson's bug description, I want to ask a clarifying question:

> Re-running pulumi up on a stack with Helm charts should not fail if the chart already exists on the cluster. Alternatively this could be a flag/option.

Did the Helm chart get installed through other means already, before running `pulumi up`? And what you're wanting is a way for the `pulumi up` to detect this, akin to a `helm upgrade --install`, instead of failing as it does today?

After rereading, particularly this part

> It would be nice to have upgrade/force as options in Pulumi helm to mimic this behavior.

I am suspecting this is what you meant, rather than a naming collision due to a bug in Pulumi, etc.

If yes, this definitely seems like a reasonable thing to support, perhaps with an `upgrade?: boolean` option that can be passed to the Helm Chart constructor.

**Comment by carterwilliamson** [comment]
I was able to reproduce with the following steps:
* Boot up AKS
* Pulumi.helm(prometheus-operator, kube-prometheus, grafana)
* Fails because prometheus-operator needs to be installed first
```
Diagnostics:
  kubernetes:monitoring.coreos.com:PrometheusRule: kube-prometheus-exporter-kube-controller-manager
    error: Plan apply failed: unable to fetch resource description for monitoring.coreos.com/v1: the server could not find the requested resource

  kubernetes:monitoring.coreos.com:ServiceMonitor: kube-prometheus-exporter-kube-state
    error: Plan apply failed: unable to fetch resource description for monitoring.coreos.com/v1alpha1: the server could not find the requested resource

  kubernetes:monitoring.coreos.com:ServiceMonitor: kube-prometheus-alertmanager
    error: Plan apply failed: unable to fetch resource description for monitoring.coreos.com/v1alpha1: the server could not find the requested resource

  kubernetes:monitoring.coreos.com:ServiceMonitor: kube-prometheus-exporter-kube-etcd
    error: Plan apply failed: unable to fetch resource description for monitoring.coreos.com/v1alpha1: the server could not find the requested resource

  kubernetes:monitoring.coreos.com:ServiceMonitor: kube-prometheus-exporter-kube-scheduler
    error: Plan apply failed: unable to fetch resource description for monitoring.coreos.com/v1alpha1: the server could not find the requested resource
```
* Add `dependsOn: prometheus-operator` to grafana and kube-prometheus
* Run `pulumi up`
* This now fails because of existing k8s resources. (See output below)
```
Diagnostics:
  kubernetes:batch:Job: prometheus-operator-get-crd
    error: Plan apply failed: 1 error occurred:

    * jobs.batch "prometheus-operator-get-crd" already exists

  kubernetes:rbac.authorization.k8s.io:ClusterRoleBinding: psp-prometheus-operator
    error: Plan apply failed: 1 error occurred:

    * clusterrolebindings.rbac.authorization.k8s.io "psp-prometheus-operator" already exists

  kubernetes:extensions:PodSecurityPolicy: prometheus-operator
    error: Plan apply failed: 1 error occurred:

    * podsecuritypolicies.extensions "prometheus-operator" already exists

  kubernetes:batch:Job: prometheus-operator-create-sm-job
    error: Plan apply failed: 1 error occurred:

    * jobs.batch "prometheus-operator-create-sm-job" already exists

  kubernetes:core:ConfigMap: prometheus-operator
    error: Plan apply failed: 1 error occurred:

    * configmaps "prometheus-operator" already exists

  kubernetes:rbac.authorization.k8s.io:ClusterRoleBinding: prometheus-operator
    error: Plan apply failed: 1 error occurred:

    * clusterrolebindings.rbac.authorization.k8s.io "prometheus-operator" already exists

  kubernetes:core:ServiceAccount: prometheus-operator
    error: Plan apply failed: 1 error occurred:

    * serviceaccounts "prometheus-operator" already exists

  kubernetes:rbac.authorization.k8s.io:ClusterRole: psp-prometheus-operator
    error: Plan apply failed: 1 error occurred:

    * clusterroles.rbac.authorization.k8s.io "psp-prometheus-operator" already exists
```

**Comment by hausdorff** [comment]
Talking to @carterwilliamson it looks like we can't reproduce this, even on his machine. We're going to watch closely in the next few days and if we catch it again I'll get into the cluster and take a look.

**Comment by carterwilliamson** [comment]
Spitballing here but I think this may be a symptom of a different problem.
Installing prometheus-operator succeeds 
```
let operator = new helm.v2.Chart("prometheus-operator", {
  repo: "coreos",
  chart: "prometheus-operator",
  version: config.operator.version,
}, { providers: { kubernetes: provider } })
```

Then followed with installing kube-prometheus fails with the error below
```
let kubePrometheus = new helm.v2.Chart("kube-prometheus", {
  repo: "coreos",
  chart: "kube-prometheus",
  version: config.kube.version,
}, { providers: { kubernetes: provider }, dependsOn: operator });
```

```
kubernetes:monitoring.coreos.com:ServiceMonitor: kube-prometheus
    error: Plan apply failed: unable to fetch resource description for monitoring.coreos.com/v1alpha1: the server could not find the requested resource
```
My guess is that this failed deployment causes the subsequent run to throw the errors related to pre-existing resources.

**Comment by hausdorff** [comment]
Hmm, I don't think that should result in the output above. That we're scheduling a create, _not_ marking that down in the checkpoint file, scheduling _another_ create, and then failing.

The only way I have ever seen this happen is if Pulumi gets interrupted catastrophically before the API server has time to report that creation succeeded.

Certainly something else could be happening here, but I very seriously doubt it's a naming problem, and I very seriously doubt that it's the problem above. This "unable to fetch resource" error happens during `Check`, before the engine even has a chance to schedule _any_ operation, let alone a create.

In short, something seriously fishy is going on around here...

**Comment by hausdorff** [comment]
Ok, checking back in with @carterwilliamson it looks like this has not resurfaced. Given the extensive attention paid to this issue, I'm comfortable closing for now. Please feel free to re-open if you see it again -- very hopefully with both the plan and application output, as this will give us some important clues about the bug. :)

**Comment by cleverguy25** [comment]
I am seeing this exact error. how can we reopen this... there needs to be a way to reconcile these when state gets out of sync.. or if someone manually does something on the cluster...

**Comment by hausdorff** [comment]
As I mentioned in my last comment, in order to be able to make progress on this issue, I don't think we're going to be able to make progress without code, the plan, the application output, and the sequence of events that led you to this situation. We tried for several hours to reproduce this and could not.

I'm not sure what you mean when you say that there needs to be a way to reconcile -- this is what `pulumi refresh` is supposed to do, no?

**Comment by cleverguy25** [comment]
I can certainly work with you to share code, plan, app output, sequence...

I have run `pulumi refresh` but it did not work...

**Comment by hausdorff** [comment]
Just to clarify: do you mean that when you ran `pulumi refresh`, we did not induct the CRDs that had been deployed previously?

**Comment by hausdorff** [comment]
To recap the offline discussion, there are several things going on:

* We were able to resolve the `prometheus-operator` issue (identical to those described here: https://github.com/pulumi/pulumi-kubernetes/issues/456). Essentially: setting `createCustomResources: false` in the values tells the chart to (correctly) not create CRDs for Prometheus, which lets the operator do that itself. It is *not* necessary to create the CRDs manually, for the same reason.

The issue @cleverguy25 is describing is actually that the semantics of `refresh` and `update` are surprising to him:

* `refresh` will fail when there is a partial failure (_e.g._, `Deployment` has not completed initialization), and the way to avoid this (using `skipAwait` in the annotations) is (a) hidden and (b) cumbersome when dealing with charts.
* `update` will fail uniformly if a resource already exists, which is not what one would expect in Kubernetes.

The former, I believe, is solved, but we'll follow up to make sure when the team confirms. The second two deserve their own issues and discussions, and I will therefore open issues for both. I will have to do it later tonight, as I won't be at my keyboard in a few minutes.

**Comment by cleverguy25** [comment]
Not exactly right for the first issue, although I believe it is a helm problem not a pulumi problem.  This is the code we had to do for prometheus-operator:

**Comment by hausdorff** [comment]
[@cleverguy25 I edited your post to put your code in a code block -- hope you don't mind]

Totally possible I could be missing something here, but when we consulted the pod logs for the prometheus operator, we found that the operator was indeed creating the CRDs on your behalf, even when `createCustomResource: false`. Testing just now, when I remove the lines that manually create the CRDs and deploy to a fresh GKE cluster, the chart works, and the CRDs get installed anyway.

The [docs] indicate that this option exists because Helm (at least some versions of it) cannot guarantee the CRDs are created before the CRs that use them. Pulumi does not have this problem because it will retry resource operations that fail.

[docs]: https://github.com/helm/charts/tree/master/stable/prometheus-operator/#helm-fails-to-create-crds
